tests/utils/test_file_utils.py

```diff
@@ -364,3 +371,106 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
+        calls_kwargs.append(kwargs)
+        response_mock = Response()
+        response_mock.status_code = 206
+        response_mock._content = b"\x01\x01"
+        return response_mock
+
+    # Workaround for windows preventing re-opening of an open handle
+    f = tempfile.NamedTemporaryFile(delete=False)
+    f.close()
+    with mock.patch("requests.Session.request") as request_mock:
+        request_mock.side_effect = mock_request_side_effect
+        parallelized_download_file_using_http_uri(
+            "fake_uri",
+            f.name,
+            file_size=1000,
+            uri_type=ArtifactCredentialType.AWS_PRESIGNED_URL,
+            chunk_size=100,
+            headers={},
+        )
+    f.close()
+    requested_ranges = [call_kwargs["headers"]["Range"] for call_kwargs in calls_kwargs]
+    assert len(requested_ranges) == 10
+    expected_ranges = []
+    for i in range(10):
+        byte_range = f"bytes={100*i}-"
+        if i != 9:
+            byte_range += f"{(100*(i+1))-1}"
+        expected_ranges.append(byte_range)
+    assert set(requested_ranges) == set(expected_ranges)
+    os.unlink(f.name)
+
+
+def test_parallelized_download_file_using_http_uri_handles_gcp_transcoding():
+    calls_kwargs = multiprocessing.Manager().list([])
+    file_content = b"\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01"
+
+    def mock_request_side_effect(method, url, *args, **kwargs):
+        calls_kwargs.append(kwargs)
+        response_mock = Response()
+        response_mock.status_code = 200
+        # 10-byte file
+        response_mock._content = file_content
+        return response_mock
+
+    f = tempfile.NamedTemporaryFile(delete=False)
+    f.close()
+    with mock.patch("requests.Session.request") as request_mock:
+        request_mock.side_effect = mock_request_side_effect
+        parallelized_download_file_using_http_uri(
+            "fake_uri",
+            f.name,
+            file_size=10,
+            uri_type=ArtifactCredentialType.GCP_SIGNED_URL,
+            chunk_size=2,
+            headers={},
+        )
+    # Should only have called once because the whole file was returned
+    assert len(calls_kwargs) == 1
+    f.seek(0)
+    contents = f.read()
+    assert contents == file_content
+    f.close()
+    os.unlink(f.name)
+
+
+def test_parallelized_download_file_using_http_uri_returns_errors_correctly():
+    calls_kwargs = multiprocessing.Manager().list([])
+    file_content = b"\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01"
+
+    def mock_request_side_effect(method, url, *args, **kwargs):
+        calls_kwargs.append(kwargs)
+        response_mock = Response()
+        response_mock.status_code = 200
+        # 10-byte file
+        response_mock._content = file_content
+        # Randomly fail for the first chunk
+        if kwargs["headers"]["Range"].startswith("bytes=0"):
+            raise requests.HTTPError("test exception")
+        return response_mock
+
+    f = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
Can we set `tmp_path` and remove `os.unlink(f.name)`?

```suggestion
    tempfile.NamedTemporaryFile(delete=False, dir=tmp_path):
```

Reply:
Could you elaborate on this fix? I think we still need `f` to be able to pass `f.name` as the download path into the function

Reply:
By supplying [`tmp_path`](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmp-path-fixture) to `dir`, the temporary file is created in a temporary directory created by pytest. pytest automatically removes this directory, so we don't need to.

Reply:
btw do we need to use `NamedTemporaryFile`? Can we just pass `tmp_path / "test"` to `download_path`?

```python
parallelized_download_file_using_http_uri("fake_uri", download_path=tmp_path / "test", ...)
```

Reply:
Got it, thanks for the tip!